R/utils.R

```diff
@@ -170,19 +174,56 @@ CGMS2DayByDay <- function(data, dt0 = NULL, inter_gap = 45, tz = ""){
   ### Interpolate on the ideal grid
   new <- as.data.frame(stats::approx(x = tr, y = g, xout = time_out))
 
-  ### Adjust to that there is no interpolation between values > inter_gap appart
-  ### Thanks to weather_interp function from weathercan R package for inspiration
-  inter_gap <- lubridate::minutes(inter_gap)
-  timediff <- lubridate::minutes(round(timediff))
-  which_gap <- tr[c(timediff > inter_gap, FALSE)]
-  missing <- lubridate::interval(which_gap + 1, which_gap + timediff[timediff > inter_gap] - 1)
-  missing <- vapply(new$x, FUN.VALUE = TRUE, FUN = function(x, missing) {
-    any(lubridate::`%within%`(x, missing))
-  }, missing = missing)
-  new$y[missing] <- NA
+  ### Adjust so that there is no interpolation between values > inter_gap apart
+  # first determine if such gaps exist and set the start+end points of such gaps to gap = TRUE
+  gaps <- tibble::tibble(
+    time = tr,
+    timediff = c(as.numeric(timediff), 0),
+    gap = (timediff > inter_gap) | (dplyr::lag(timediff, 1L, default = FALSE) > inter_gap)
+  )
+  # if any gaps exist, set any interpolated values in those times to NA
+  if (any(gaps$gap)) {
+    missing <- gaps %>%
+      dplyr::filter(gap) %>%
+      dplyr::mutate(
+        # starts have timediff greater than inter_gap
+        start = dplyr::if_else(timediff > inter_gap, time + 1, NA_real_),
+        # ends are the ones where the time before is the gap start
+        end = dplyr::if_else(dplyr::lag(timediff, 1L, default = FALSE) > inter_gap,
+                             time - 1, NA_real_),
+        # shift ends to match with their gap start time point
+        end = c(end[2:dplyr::n()], NA)
+      ) %>%
+      dplyr::filter(!is.na(start)) # filter down to one entry per gap
+
+    # create grid of interpolated time points that exist in each gap
+    # rule = 1 means points outside of xmin/xmax will be NA (no extrapolation)
+    intervals <- list(nrow(missing))
+    for (i in 1:nrow(missing)) {
+      grid = stats::approx(x = c(missing$start[i], missing$end[i]),
```

Comment:
I am a little confused why you interpolate again. Line 175 above did all interpolation already (including over missing regions). So after the gaps are identified, new$y should be set to NA within gaps (no more interpolation).

Reply:
if my understanding is correct, then once you identified start/end times, you should be setting new$y directly to NA for elements corresponding to those times (new$x > start time and new$x < end time), does this make sense?